neurom/core/morphology.py

```diff
@@ -240,23 +259,45 @@ def iter_neurites(obj, mapfun=None, filt=None, neurite_order=NeuriteIter.FileOrd
         >>> mapping = lambda n : len(n.points)
         >>> n_points = [n for n in iter_neurites(pop, mapping, filter)]
     """
-    neurites = ((obj,) if isinstance(obj, Neurite) else
-                obj.neurites if hasattr(obj, 'neurites') else obj)
+    def extract_subneurites(neurite):
+        if neurite.is_heterogeneous():
+            return _homogeneous_subtrees(neurite)
+        return [neurite]
+
+    neurites = (
+        (obj,)
+        if isinstance(obj, Neurite)
+        else obj.neurites
+        if hasattr(obj, "neurites")
+        else obj
+    )
+
     if neurite_order == NeuriteIter.NRN:
         if isinstance(obj, Population):
             warnings.warn('`iter_neurites` with `neurite_order` over Population orders neurites'
                           'within the whole population, not within each morphology separately.')
         last_position = max(NRN_ORDER.values()) + 1
         neurites = sorted(neurites, key=lambda neurite: NRN_ORDER.get(neurite.type, last_position))
 
+    if use_subtrees:
+        neurites = flatten(map(extract_subneurites, neurites))
+
     neurite_iter = iter(neurites) if filt is None else filter(filt, neurites)
-    return neurite_iter if mapfun is None else map(mapfun, neurite_iter)
+
+    if mapfun is None:
+        return neurite_iter
+
+    if use_subtrees:
+        return (mapfun(neurite, section_type=neurite.type) for neurite in neurite_iter)
```

Comment:
```suggestion
        mapfun = functools.partial(mapfun, section_type=neurite.type)
```

Reply:
I'd also move that into the `if use_subtrees` (w/ a guard on `mapfun is not None`)

Reply:
The main issue with this is that it passes the type of each neurite, not a fixed one (neurite.type vs. neurite_type).